File: backend/src/rank.js
function parseRank(rankStr) {
  if (typeof rankStr !== 'string') return null;

  const gupMatch = rankStr.match(/^([1-9]|1[0-8])급$/);
  if (gupMatch) {
    return { type: 'gup', value: Number(gupMatch[1]) };
  }

  const danMatch = rankStr.match(/^([1-9])단$/);
  if (danMatch) {
    return { type: 'dan', value: Number(danMatch[1]) };
  }

  return null;
}

function rankToTier(rankStr) {
  const parsed = parseRank(rankStr);
  if (!parsed) return 0;

  if (parsed.type === 'gup') {
    // 18급 -> 0, 1급 -> 17
    return 18 - parsed.value;
  }

  // 1단 -> 18, 9단 -> 26
  return 17 + parsed.value;
}

function tierToRank(tier) {
  const normalized = Math.max(0, Math.min(26, tier));
  if (normalized <= 17) {
    return `${18 - normalized}급`;
  }
  return `${normalized - 17}단`;
}

function getRankThreshold(rankStr) {
  const parsed = parseRank(rankStr) || { type: 'gup', value: 18 };
  if (parsed.type === 'dan') return 7;
  // 18급 ~ 10급
  if (parsed.value >= 10) return 3;
  // 9급 ~ 1급
  return 5;
}

function canPromote(rankStr) {
  return rankToTier(rankStr) < 26;
}

function canDemote(rankStr) {
  return rankToTier(rankStr) > 0;
}

function promoteRank(rankStr) {
  return tierToRank(rankToTier(rankStr) + 1);
}

function demoteRank(rankStr) {
  return tierToRank(rankToTier(rankStr) - 1);
}

function normalizeCounter(value) {
  return Math.max(0, Number.isFinite(Number(value)) ? Number(value) : 0);
}

function resolveRankAfterResult(rankStr, rankWins, rankLosses, result) {
  let nextRank = parseRank(rankStr) ? rankStr : '18급';
  let wins = normalizeCounter(rankWins);
  let losses = normalizeCounter(rankLosses);

  if (result === 'win') wins += 1;
  if (result === 'loss') losses += 1;

  // At most one rank change per game.
  const threshold = getRankThreshold(nextRank);
  if (wins >= threshold && canPromote(nextRank)) {
    nextRank = promoteRank(nextRank);
    wins = 0;
    losses = 0;
  } else if (losses >= threshold && canDemote(nextRank)) {
    nextRank = demoteRank(nextRank);
    wins = 0;
    losses = 0;
  } else {
    // Bound counters when rank cannot move in that direction.
    if (!canPromote(nextRank)) wins = Math.min(wins, threshold);
    if (!canDemote(nextRank)) losses = Math.min(losses, threshold);
  }

  return {
    rank: nextRank,
    rankWins: wins,
    rankLosses: losses,
  };
}

module.exports = {
  parseRank,
  rankToTier,
  tierToRank,
  getRankThreshold,
  canPromote,
  canDemote,
  promoteRank,
  demoteRank,
  normalizeCounter,
  resolveRankAfterResult,
};

File: backend/rank.go
package main

import "strings"

type parsedRank struct {
	Type  string
	Value int
}

type rankState struct {
	Rank       string `json:"rank"`
	RankWins   int    `json:"rankWins"`
	RankLosses int    `json:"rankLosses"`
}

func parseRank(rank string) *parsedRank {
	if strings.HasSuffix(rank, gupSuffix) {
		value, ok := toInt(strings.TrimSuffix(rank, gupSuffix))
		if ok && value >= 1 && value <= 18 {
			return &parsedRank{Type: "gup", Value: value}
		}
	}
	if strings.HasSuffix(rank, danSuffix) {
		value, ok := toInt(strings.TrimSuffix(rank, danSuffix))
		if ok && value >= 1 && value <= 9 {
			return &parsedRank{Type: "dan", Value: value}
		}
	}
	return nil
}

func rankToTier(rank string) int {
	parsed := parseRank(rank)
	if parsed == nil {
		return 0
	}
	if parsed.Type == "gup" {
		return 18 - parsed.Value
	}
	return 17 + parsed.Value
}

func tierToRank(tier int) string {
	if tier < 0 {
		tier = 0
	}
	if tier > 26 {
		tier = 26
	}
	if tier <= 17 {
		return itoa(18-tier) + gupSuffix
	}
	return itoa(tier-17) + danSuffix
}

func getRankThreshold(rank string) int {
	parsed := parseRank(rank)
	if parsed == nil {
		return 3
	}
	if parsed.Type == "dan" {
		return 7
	}
	if parsed.Value >= 10 {
		return 3
	}
	return 5
}

func canPromote(rank string) bool {
	return rankToTier(rank) < 26
}

func canDemote(rank string) bool {
	return rankToTier(rank) > 0
}

func promoteRank(rank string) string {
	return tierToRank(rankToTier(rank) + 1)
}

func demoteRank(rank string) string {
	return tierToRank(rankToTier(rank) - 1)
}

func normalizeCounter(value any) int {
	parsed, ok := toInt(value)
	if !ok || parsed < 0 {
		return 0
	}
	return parsed
}

func resolveRankAfterResult(rank string, rankWins, rankLosses any, result string) rankState {
	nextRank := rank
	if parseRank(nextRank) == nil {
		nextRank = "18" + gupSuffix
	}

	wins := normalizeCounter(rankWins)
	losses := normalizeCounter(rankLosses)

	if result == "win" {
		wins++
	}
	if result == "loss" {
		losses++
	}

	threshold := getRankThreshold(nextRank)
	if wins >= threshold && canPromote(nextRank) {
		nextRank = promoteRank(nextRank)
		wins = 0
		losses = 0
	} else if losses >= threshold && canDemote(nextRank) {
		nextRank = demoteRank(nextRank)
		wins = 0
		losses = 0
	} else {
		if !canPromote(nextRank) && wins > threshold {
			wins = threshold
		}
		if !canDemote(nextRank) && losses > threshold {
			losses = threshold
		}
	}

	return rankState{
		Rank:       nextRank,
		RankWins:   wins,
		RankLosses: losses,
	}
}
